endpoints/openrtb2/amp_auction.go

```diff
@@ -148,8 +148,17 @@ func (deps *endpointDeps) AmpAuction(w http.ResponseWriter, r *http.Request, _ h
 	} else {
 		labels.CookieFlag = pbsmetrics.CookieFlagYes
 	}
-	if req.Site != nil && req.Site.Publisher != nil && req.Site.Publisher.ID != "" {
-		labels.PubID = req.Site.Publisher.ID
+	if req.Site != nil && req.Site.Publisher != nil {
+		if req.Site.Publisher.ID != "" {
+			labels.PubID = req.Site.Publisher.ID
+		}
+		var pubExt openrtb_ext.ExtPublisher
+		if req.Site.Ext != nil {
+			err := json.Unmarshal(req.Site.Ext, &pubExt)
+			if err == nil && pubExt.ParentAccount != nil {
+				labels.PubID = *pubExt.ParentAccount
+			}
+		}
```

Comment:
It seems that in almost every changed file in this request, the ordering of checks can be changed in order to make the code more efficient.  For example, in the above  code, we go to the trouble of checking req.Site.Publisher and req.Site.Publisher.ID and assigning labels.PubID only to reassign it later if req.Site.Ext can be unmarshalled to successfully obtain ParentAccount.  It would be more efficient to first try to get PubID from req.Site.Ext and look at req.Site.Publisher only if necessary.

It's not clear to me whether we want to not use req.Site.Ext in the case that req.Site.Publisher is nil.  I would guess that's not the case  (i.e., if req.Site.Publisher is nil and we can get PubID from req.Site.Ext, I expect we want to get it), though that is what the code currently does.  If I'm wrong, that would slightly alter my above comment.

Reply:
Oh, you caught a bug. It should be `req.Site.Publisher.Ext`. I am a little up in the air if grabbing `PubID` from `pubExt` first and then checking if we need to try `Publisher.ID` will save anything. It saves an assignment from one variable to another, but adds a check that will always be needed (Did we successfully populate from `pubExt`?). I suppose we could duplicate the code to appear on both `else` branches of the decision tree to avoid the check.